intelmq/lib/bot.py

```diff
@@ -60,6 +61,8 @@ class Bot(object):
     is_multithreadable = True
     # Collectors with an empty process() should set this to true, prevents endless loops (#1364)
     collector_empty_process = False
+    argparser = argparse.ArgumentParser(usage='%(prog)s [OPTIONS] BOT-ID')
```

Comment:
I did think about it and decided to place it outside the constructor for the following reasons:
* The content of ArgumentParser (subparsers, arguments, etc) is the same for all the instances of the bot which puts it in the position of class variable rather than instance variable. The ArgumentParser doesn't need to be instantiated with every bot instance.
* The main `run()` function is class method and the bot is only instantiated at the end of it therefore there wouldn't be available any ArgumentParser instance during the `run()` function when we need to parse the arguments.
* The `update_database` function was specifically built as a class method to be used in a fashion of a standalone script. It doesn't need to have an instance of the bot.

Now the real question is whether it is **desired to have any extra bot functionality in form of class methods** (== we do not need an instance of the bot) **or in the form of instance methods**. Both have it's pros and cons.

**Pros** (of having extra bot functionality as **instance methods**):
* Bot parameters and all of it's configuration is loaded and handled by the underlying class, you don't need to load and validate anything.

**Cons** (of having extra bot functionality as **instance methods**):
* If the bot is not fully configured it might fail when trying to instantiate the bot. For example if `update_database` was an instance method, it wouldn't be possible to call it without having correctly configured `pipeline.conf`, because `lib.bot.__init__` would fail on `lib.bot.__load_pipeline_configuration` which is rather silly considering that the pipeline has nothing to do with updating the database.
* You need to specify the bot id (to have the bot's configuration loaded), which is something that you might not want to do (or can't). This is supposed to be one of the benefits of this PR. Another example I can think of is (my) desire to use the `experts.sieve` to validate a filter syntax without the need to have the bot properly configured.

**Pros** (of having extra bot functionality as **class methods**):
* Removes cons of having extra functionality as instance methods (above).

**Cons** (of having extra bot functionality as **class methods**):
* Some code needs to be more or less duplicated (or perhaps have one code usable in both ways).
* Specific bot configuration might be ignored (in very exotic cases) because of missing bot-id. For example with `update_database` if we had two `experts.asn_lookup` bots with different proxy settings for some reason, the function has no clue which one to pick (currently it is completely ignored and the `defaults.conf` is used).

After considering all this I decided for this PR that having class method is the better approach. My long comment is also relevant to the next conversation regarding the code duplication in `def create_request_session() -> requests.Session:`. 

I welcome any insights on this.

Reply:
Thanks for your thorough response, I did by far not expect such a detailed explanation on such a short comment.

tl;dr I'm totally ok with leaving it there.

>     * The content of ArgumentParser (subparsers, arguments, etc) is the same for all the instances of the bot which puts it in the position of class variable rather than instance variable. The ArgumentParser doesn't need to be instantiated with every bot instance.

I'm not sure if that matters somewhere (I.e. we create a bot instance and the constructor is not called).

> Now the real question is whether it is **desired to have any extra bot functionality in form of class methods** (== we do not need an instance of the bot) **or in the form of instance methods**. Both have it's pros and cons.

We may end up with both. We now have class methods because it is easier or at all possible (for instance methods we block the actually running bot instance)

>     * Some code needs to be more or less duplicated (or perhaps have one code usable in both ways).

I prefer the latter

>     * Specific bot configuration might be ignored (in very exotic cases) because of missing bot-id. For example with `update_database` if we had two `experts.asn_lookup` bots with different proxy settings for some reason, the function has no clue which one to pick (currently it is completely ignored and the `defaults.conf` is used).

This is an issue, but can be solved by (more) code.

I agree with everything else.